SmallRyeOpenApiProcessor throw exception on org.javatuples.Pair

Add these two dependencies to pom.xml of a Quarkus project:

```
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-smallrye-openapi</artifactId>
        </dependency>

        <dependency>
            <groupId>com.flipkart.utils</groupId>
            <artifactId>javatuples</artifactId>
            <version>3.0</version>
        </dependency>
```

Add a JAX-RS resource class:
```
import org.javatuples.Pair;

@Path("/hello")
public class HelloResource {
  @GET
  public Pair<String, String> hello() {
    return Pair.with("hello", "world");
  }
}
```

Run unit test `HelloResourceTest` will fail:

```package com.yingmi.scrm.user.mapping;

import io.quarkus.test.common.http.TestHTTPEndpoint;
import io.quarkus.test.junit.QuarkusTest;
import org.junit.jupiter.api.Test;

@QuarkusTest
@TestHTTPEndpoint(UserBindingResourceTest.class)
public class UserBindingResourceTest {

    @Test
    public void testAdd() {

    }
}
```

Test result:
```
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.IllegalArgumentException: Not a parameterized type!
	at org.jboss.jandex.Type.asParameterizedType(Type.java:228)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.createParameterizedType(TypeResolver.java:1088)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.resolveParameterizedAncestor(TypeResolver.java:1073)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.resolveSpecial(OpenApiDataObjectScanner.java:330)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:208)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:159)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.otherTypeToSchema(SchemaFactory.java:610)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.typeToSchema(SchemaFactory.java:448)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.createResponseFromRestMethod(AnnotationScanner.java:423)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.processResponse(AnnotationScanner.java:354)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:441)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:277)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:275)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:245)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:220)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:152)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:705)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:563)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: java.lang.RuntimeException: 
io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.IllegalArgumentException: Not a parameterized type!
	at org.jboss.jandex.Type.asParameterizedType(Type.java:228)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.createParameterizedType(TypeResolver.java:1088)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.resolveParameterizedAncestor(TypeResolver.java:1073)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.resolveSpecial(OpenApiDataObjectScanner.java:330)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:208)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:159)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.otherTypeToSchema(SchemaFactory.java:610)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.typeToSchema(SchemaFactory.java:448)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.createResponseFromRestMethod(AnnotationScanner.java:423)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.processResponse(AnnotationScanner.java:354)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:441)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:277)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:275)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:245)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:220)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:152)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:705)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:563)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: io.quarkus.builder.BuildException: 
Build failure: Build failed due to errors
	[error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.IllegalArgumentException: Not a parameterized type!
	at org.jboss.jandex.Type.asParameterizedType(Type.java:228)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.createParameterizedType(TypeResolver.java:1088)
	at io.smallrye.openapi.runtime.scanner.dataobject.TypeResolver.resolveParameterizedAncestor(TypeResolver.java:1073)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.resolveSpecial(OpenApiDataObjectScanner.java:330)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:208)
	at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:159)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.otherTypeToSchema(SchemaFactory.java:610)
	at io.smallrye.openapi.runtime.io.schema.SchemaFactory.typeToSchema(SchemaFactory.java:448)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.createResponseFromRestMethod(AnnotationScanner.java:423)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.processResponse(AnnotationScanner.java:354)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:441)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:277)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:275)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:245)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:220)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:152)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:705)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:563)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

Caused by: java.lang.IllegalArgumentException: Not a parameterized type!
```

--- a/core/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeResolver.java
+++ b/core/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeResolver.java
@@ -1090,6 +1090,10 @@ public class TypeResolver {
     }
 
     private static ParameterizedType createParameterizedType(Type targetType, Map<String, Type> resolutionMap) {
+        if (targetType.kind() != Type.Kind.PARAMETERIZED_TYPE) {
+            return ParameterizedType.create(targetType.name(), Type.EMPTY_ARRAY, null);
+        }
+
         Type[] resolvedArgs = resolveArguments(targetType.asParameterizedType(), t -> resolveType(t, resolutionMap));
         return ParameterizedType.create(targetType.name(), resolvedArgs, null);
     }

--- a/core/src/main/java/io/smallrye/openapi/runtime/util/TypeUtil.java
+++ b/core/src/main/java/io/smallrye/openapi/runtime/util/TypeUtil.java
@@ -487,6 +487,12 @@ public class TypeUtil {
             case WILDCARD_TYPE:
                 terminal = false;
                 break;
+            case CLASS:
+                if (DOTNAME_OBJECT.equals(type.name())) {
+                    terminal = true;
+                    break;
+                }
+                // Fall through
             default:
                 // If is known type.
                 terminal = !getTypeFormat(type).isSchemaType(SchemaType.ARRAY, SchemaType.OBJECT);

--- a/core/src/test/java/io/smallrye/openapi/runtime/scanner/StandaloneSchemaScanTest.java
+++ b/core/src/test/java/io/smallrye/openapi/runtime/scanner/StandaloneSchemaScanTest.java
@@ -7,6 +7,7 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -510,5 +511,33 @@ class StandaloneSchemaScanTest extends IndexScannerTestBase {
         OpenAPI result = scanner.scan();
         printToConsole(result);
         assertJsonEquals("components.schemas.annotated-constructor-arg-ignored.json", result);
+    }
+
+    /*
+     * https://github.com/smallrye/smallrye-open-api/issues/944
+     */
+    @Test
+    void testParameterizedTypeWithNonparameterizedAncestryChainLink() throws IOException, JSONException {
+        class Tuple implements Iterable<Object> {
+            @Override
+            public Iterator<Object> iterator() {
+                return null;
+            }
+        }
+
+        class Pair<T1, T2> extends Tuple {
+        }
+
+        @Schema(name = "TestBean")
+        class Bean {
+            @SuppressWarnings("unused")
+            Pair<String, String> pair;
+        }
+
+        Index index = indexOf(Bean.class, Pair.class, Tuple.class);
+        OpenApiAnnotationScanner scanner = new OpenApiAnnotationScanner(emptyConfig(), index);
+        OpenAPI result = scanner.scan();
+        printToConsole(result);
+        assertJsonEquals("components.schemas.nonparameterized-ancestry-chain-link.json", result);
     }
 }

--- a/core/src/test/resources/io/smallrye/openapi/runtime/scanner/components.schemas.nonparameterized-ancestry-chain-link.json
+++ b/core/src/test/resources/io/smallrye/openapi/runtime/scanner/components.schemas.nonparameterized-ancestry-chain-link.json
@@ -1,0 +1,21 @@
+{
+  "openapi": "3.0.3",
+  "components": {
+    "schemas": {
+      "1PairStringString": {
+        "type": "array",
+        "items": {
+          "type": "object"
+        }
+      },
+      "TestBean": {
+        "type": "object",
+        "properties": {
+          "pair": {
+            "$ref": "#/components/schemas/1PairStringString"
+          }
+        }
+      }
+    }
+  }
+}

